rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java

```diff
@@ -306,27 +316,106 @@ protected TableRecords queryCurrentRecords(Connection conn) throws SQLException
         return currentRecords;
     }
 
+    protected List<Field> getOrderedPkList(TableRecords image,Row row,String dbType) {
+        List<Field> pkFields = new ArrayList<>();
+        // To ensure the order of the pk, the order should based on getPrimaryKeyOnlyName.
+        List<String> pkColumnNameListByOrder = image.getTableMeta().getPrimaryKeyOnlyName();
+        List<String> pkColumnNameListNoOrder = row.primaryKeys()
+                .stream()
+                .map(e -> ColumnUtils.delEscape(e.getName(),dbType))
+                .collect(Collectors.toList());
+        pkColumnNameListByOrder.forEach(pkName -> {
+            int pkIndex = pkColumnNameListNoOrder.indexOf(pkName);
+            if (pkIndex != -1) {
+                // add PK to the last of the list.
+                pkFields.add(row.primaryKeys().get(pkIndex));
+            }
+        });
+        return pkFields;
+    }
+
+    /**
+     * each pk is a condition.the result will like :" id =? and userCode =?"
+     * @param pkNameList
+     * @return return where condition sql string.the sql can just search one related record.
+     */
+    protected String buildWhereConditionByPKs(List<String> pkNameList,KeywordChecker keywordChecker) {
+        StringBuilder whereStr = new StringBuilder();
+        //we must consider the situation of composite primary key
+        for (int i = 0;i < pkNameList.size(); i++) {
+            if (i > 0) {
+                whereStr.append(" and ");
+            }
+            String pkName = pkNameList.get(i);
+            whereStr.append(keywordChecker.checkAndReplace(pkName));
+            whereStr.append(" = ? ");
+        }
+        return whereStr.toString();
+
+    }
+    /**
+     * each pk is a condition.the result will like :" id in (?,?,?) and userCode in (?,?,?)"
+     * @param pkNameList
+     * @param pkRowValues  the kye of map is pk name ,and value of map is pk's value
+     * @return return where condition sql string.the sql can search all related records not just one.
+     */
+    protected String buildWhereConditionByPKs(List<String> pkNameList,Map<String,List<Field>> pkRowValues,Connection conn) throws SQLException {
```

Comment:
(a1 and b1) or ( a1 and b2 ) or (a2 and b2)  <=>  A in (a1,a2) and B in (b1,b2) ?

Reply:
ok